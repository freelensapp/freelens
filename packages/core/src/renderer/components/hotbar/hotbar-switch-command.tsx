/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import { withInjectables } from "@ogre-tools/injectable-react";
import type { IComputedValue } from "mobx";
import { observer } from "mobx-react";
import React from "react";
import type { ComputeHotbarDisplayLabel } from "../../../features/hotbar/storage/common/compute-display-label.injectable";
import computeHotbarDisplayLabelInjectable from "../../../features/hotbar/storage/common/compute-display-label.injectable";
import type { Hotbar } from "../../../features/hotbar/storage/common/hotbar";
import hotbarsInjectable from "../../../features/hotbar/storage/common/hotbars.injectable";
import type { SetAsActiveHotbar } from "../../../features/hotbar/storage/common/set-as-active.injectable";
import setAsActiveHotbarInjectable from "../../../features/hotbar/storage/common/set-as-active.injectable";
import type { CommandOverlay } from "../command-palette";
import commandOverlayInjectable from "../command-palette/command-overlay.injectable";
import { Select } from "../select";
import { HotbarAddCommand } from "./hotbar-add-command";
import { HotbarRemoveCommand } from "./hotbar-remove-command";
import { HotbarRenameCommand } from "./hotbar-rename-command";

const hotbarAddAction = Symbol("hotbar-add");
const hotbarRemoveAction = Symbol("hotbar-remove");
const hotbarRenameAction = Symbol("hotbar-rename");

interface Dependencies {
  setAsActiveHotbar: SetAsActiveHotbar;
  computeHotbarDisplayLabel: ComputeHotbarDisplayLabel;
  hotbars: IComputedValue<Hotbar[]>;
  commandOverlay: CommandOverlay;
}

function ignoreIf<T>(check: boolean, menuItems: T[]): T[] {
  return check ? [] : menuItems;
}

const NonInjectedHotbarSwitchCommand = observer(
  ({ setAsActiveHotbar, computeHotbarDisplayLabel, hotbars, commandOverlay }: Dependencies) => (
    <Select
      id="switch-to-hotbar-input"
      menuPortalTarget={null}
      onChange={(option) => {
        if (!option) {
          return;
        }

        if (typeof option.value === "symbol") {
          switch (option.value) {
            case hotbarAddAction:
              return commandOverlay.open(<HotbarAddCommand />);
            case hotbarRemoveAction:
              return commandOverlay.open(<HotbarRemoveCommand />);
            case hotbarRenameAction:
              return commandOverlay.open(<HotbarRenameCommand />);
          }
        } else {
          setAsActiveHotbar(option.value);
          commandOverlay.close();
        }
      }}
      components={{ DropdownIndicator: null, IndicatorSeparator: null }}
      menuIsOpen={true}
      options={[
        ...hotbars.get().map((hotbar) => ({
          value: hotbar,
          label: computeHotbarDisplayLabel(hotbar),
        })),
        {
          value: hotbarAddAction,
          label: "Add hotbar ...",
        },
        ...ignoreIf(hotbars.get().length <= 1, [
          {
            value: hotbarRemoveAction,
            label: "Remove hotbar ...",
          },
        ]),
        {
          value: hotbarRenameAction,
          label: "Rename hotbar ...",
        },
      ]}
      autoFocus={true}
      escapeClearsValue={false}
      isClearable={false}
      placeholder="Switch to hotbar"
    />
  ),
);

export const HotbarSwitchCommand = withInjectables<Dependencies>(NonInjectedHotbarSwitchCommand, {
  getProps: (di, props) => ({
    ...props,
    commandOverlay: di.inject(commandOverlayInjectable),
    computeHotbarDisplayLabel: di.inject(computeHotbarDisplayLabelInjectable),
    hotbars: di.inject(hotbarsInjectable),
    setAsActiveHotbar: di.inject(setAsActiveHotbarInjectable),
  }),
});
