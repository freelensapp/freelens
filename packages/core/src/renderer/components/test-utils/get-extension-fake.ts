/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import { LensMainExtension } from "../../../extensions/lens-main-extension";
import { LensRendererExtension } from "../../../extensions/lens-renderer-extension";

export class TestExtensionMain extends LensMainExtension {}
export class TestExtensionRenderer extends LensRendererExtension {}

export interface FakeExtensionOptions {
  id: string;
  name: string;
  rendererOptions?: Partial<LensRendererExtension>;
  mainOptions?: Partial<LensMainExtension>;
}

export const getExtensionFakeForMain = ({ id, name, mainOptions = {} }: FakeExtensionOptions) =>
  Object.assign(
    new TestExtensionMain({
      id,
      absolutePath: "irrelevant",
      isBundled: false,
      isCompatible: false,
      isEnabled: false,
      manifest: {
        name,
        version: "1.0.0",
        engines: {
          freelens: "^0.1.0",
        },
      },
      manifestPath: "irrelevant",
    }),
    mainOptions,
  );

export const getExtensionFakeForRenderer = ({ id, name, rendererOptions = {} }: FakeExtensionOptions) =>
  Object.assign(
    new TestExtensionRenderer({
      id,
      absolutePath: "irrelevant",
      isBundled: false,
      isCompatible: false,
      isEnabled: false,
      manifest: {
        name,
        version: "1.0.0",
        engines: {
          freelens: "^0.1.0",
        },
      },
      manifestPath: "irrelevant",
    }),
    rendererOptions,
  );
