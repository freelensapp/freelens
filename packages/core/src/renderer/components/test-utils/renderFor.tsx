/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import type { DiContainer } from "@ogre-tools/injectable";
import { DiContextProvider } from "@ogre-tools/injectable-react";
import type { RenderResult } from "@testing-library/react";
import { render as testingLibraryRender } from "@testing-library/react";
import React from "react";

export type DiRender = (ui: React.ReactElement) => RenderResult;

type DiRenderFor = (di: DiContainer) => DiRender;

export const renderFor: DiRenderFor = (di) => (ui) => {
  const result = testingLibraryRender(<DiContextProvider value={{ di }}>{ui}</DiContextProvider>);

  return {
    ...result,

    rerender: (ui: React.ReactElement) => result.rerender(<DiContextProvider value={{ di }}>{ui}</DiContextProvider>),
  };
};
