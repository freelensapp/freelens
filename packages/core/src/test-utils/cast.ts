/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

export const cast = <T>(data: Partial<T>): T => data as T;
