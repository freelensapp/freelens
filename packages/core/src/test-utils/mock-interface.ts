/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import type { AsyncFnMock } from "@async-fn/jest";

type GetMockedType<T> = T extends (...args: any[]) => Promise<any>
  ? AsyncFnMock<T>
  : T extends (...args: any[]) => any
    ? jest.MockedFunction<T>
    : T;

export type Mocked<T extends object> = {
  -readonly [P in keyof T]: GetMockedType<T[P]>;
};
