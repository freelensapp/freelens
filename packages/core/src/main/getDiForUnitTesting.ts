/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import { registerFeature } from "@freelensapp/feature-core";
import { kubeApiSpecificsFeature } from "@freelensapp/kube-api-specifics";
import { setLegacyGlobalDiForExtensionApi } from "@freelensapp/legacy-global-di";
import { loggerFeature } from "@freelensapp/logger";
import { messagingFeature, testUtils as messagingTestUtils } from "@freelensapp/messaging";
import { notificationsFeature } from "@freelensapp/notifications";
import { randomFeature } from "@freelensapp/random";
import type { GlobalOverride } from "@freelensapp/test-utils";
import type { DiContainer } from "@ogre-tools/injectable";
import { createContainer, isInjectable } from "@ogre-tools/injectable";
import { registerMobX } from "@ogre-tools/injectable-extension-for-mobx";
import { chunk } from "lodash/fp";
import { runInAction } from "mobx";
import broadcastMessageInjectable from "../common/ipc/broadcast-message.injectable";
import setupSyncingOfWeblinksInjectable from "../features/weblinks/main/setup-syncing-of-weblinks.injectable";
import { getOverrideFsWithFakes } from "../test-utils/override-fs-with-fakes";
import spawnInjectable from "./child-process/spawn.injectable";
import initializeClusterManagerInjectable from "./cluster/initialize-manager.injectable";
import electronQuitAndInstallUpdateInjectable from "./electron-app/features/electron-quit-and-install-update.injectable";
import electronUpdaterIsActiveInjectable from "./electron-app/features/electron-updater-is-active.injectable";
import setUpdateOnQuitInjectable from "./electron-app/features/set-update-on-quit.injectable";
import setupApplicationNameInjectable from "./electron-app/runnables/setup-application-name.injectable";
import setupDeepLinkingInjectable from "./electron-app/runnables/setup-deep-linking.injectable";
import setupDeviceShutdownInjectable from "./electron-app/runnables/setup-device-shutdown.injectable";
import setupIpcMainHandlersInjectable from "./electron-app/runnables/setup-ipc-main-handlers/setup-ipc-main-handlers.injectable";
import setupMainWindowVisibilityAfterActivationInjectable from "./electron-app/runnables/setup-main-window-visibility-after-activation.injectable";
import waitUntilBundledExtensionsAreLoadedInjectable from "./start-main-application/lens-window/application-window/wait-until-bundled-extensions-are-loaded.injectable";
import initializeExtensionsInjectable from "./start-main-application/runnables/initialize-extensions.injectable";
import setupLensProxyInjectable from "./start-main-application/runnables/setup-lens-proxy.injectable";

export function getDiForUnitTesting() {
  const environment = "main";
  const di = createContainer(environment, {
    detectCycles: false,
  });

  registerMobX(di);
  setLegacyGlobalDiForExtensionApi(di, environment);

  runInAction(() => {
    registerFeature(
      di,
      messagingFeature,
      messagingTestUtils.messagingFeatureForUnitTesting,
      loggerFeature,
      randomFeature,
      kubeApiSpecificsFeature,
      notificationsFeature,
    );
  });

  di.preventSideEffects();

  runInAction(() => {
    const injectables = global.injectablePaths.main.paths
      .map((path) => require(path))
      .flatMap(Object.values)
      .filter(isInjectable);

    for (const block of chunk(100)(injectables)) {
      di.register(...block);
    }
  });

  for (const globalOverridePath of global.injectablePaths.main.globalOverridePaths) {
    const globalOverride = require(globalOverridePath).default as GlobalOverride<unknown, unknown, unknown>;

    di.override(globalOverride.injectable, globalOverride.overridingInstantiate);
  }

  di.override(waitUntilBundledExtensionsAreLoadedInjectable, () => async () => {});

  overrideRunnablesHavingSideEffects(di);
  overrideElectronFeatures(di);
  getOverrideFsWithFakes()(di);

  di.override(broadcastMessageInjectable, () => (channel) => {
    throw new Error(`Tried to broadcast message to channel "${channel}" over IPC without explicit override.`);
  });
  di.override(spawnInjectable, () => () => {
    return {
      stderr: { on: jest.fn(), removeAllListeners: jest.fn() },
      stdout: { on: jest.fn(), removeAllListeners: jest.fn() },
      on: jest.fn(),
    } as never;
  });

  return di;
}

// TODO: Reorganize code in Runnables to get rid of requirement for override
const overrideRunnablesHavingSideEffects = (di: DiContainer) => {
  [
    initializeExtensionsInjectable,
    initializeClusterManagerInjectable,
    setupIpcMainHandlersInjectable,
    setupLensProxyInjectable,
    setupSyncingOfWeblinksInjectable,
  ].forEach((injectable) => {
    di.override(injectable, () => ({
      id: injectable.id,
      run: () => {},
    }));
  });
};

const overrideElectronFeatures = (di: DiContainer) => {
  [
    setupMainWindowVisibilityAfterActivationInjectable,
    setupDeviceShutdownInjectable,
    setupDeepLinkingInjectable,
    setupApplicationNameInjectable,
  ].forEach((injectable) => {
    di.override(injectable, () => ({
      id: injectable.id,
      run: () => {},
    }));
  });

  di.override(electronQuitAndInstallUpdateInjectable, () => () => {});
  di.override(setUpdateOnQuitInjectable, () => () => {});
  di.override(electronUpdaterIsActiveInjectable, () => false);
};
