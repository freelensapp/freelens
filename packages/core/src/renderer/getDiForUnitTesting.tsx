/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import { animateFeature, requestAnimationFrameInjectable } from "@freelensapp/animate";
import { clusterSidebarFeature } from "@freelensapp/cluster-sidebar";
import { registerFeature } from "@freelensapp/feature-core";
import { kubeApiSpecificsFeature } from "@freelensapp/kube-api-specifics";
import { setLegacyGlobalDiForExtensionApi } from "@freelensapp/legacy-global-di";
import { loggerFeature } from "@freelensapp/logger";
import { messagingFeature, testUtils as messagingTestUtils } from "@freelensapp/messaging";
import { notificationsFeature } from "@freelensapp/notifications";
import { randomFeature } from "@freelensapp/random";
import { routingFeature } from "@freelensapp/routing";
import type { GlobalOverride } from "@freelensapp/test-utils";
import { createContainer, isInjectable } from "@ogre-tools/injectable";
import { registerMobX } from "@ogre-tools/injectable-extension-for-mobx";
import { registerInjectableReact } from "@ogre-tools/injectable-react";
import { chunk, noop } from "lodash/fp";
import { runInAction } from "mobx";
import { getOverrideFsWithFakes } from "../test-utils/override-fs-with-fakes";
import hostedClusterIdInjectable from "./cluster-frame-context/hosted-cluster-id.injectable";
import terminalSpawningPoolInjectable from "./components/dock/terminal/terminal-spawning-pool.injectable";
import startTopbarStateSyncInjectable from "./components/layout/top-bar/start-state-sync.injectable";
import legacyOnChannelListenInjectable from "./ipc/legacy-channel-listen.injectable";
import watchHistoryStateInjectable from "./remote-helpers/watch-history-state.injectable";

export const getDiForUnitTesting = () => {
  const environment = "renderer";
  const di = createContainer(environment, {
    detectCycles: false,
  });

  registerMobX(di);
  registerInjectableReact(di);
  setLegacyGlobalDiForExtensionApi(di, environment);

  runInAction(() => {
    registerFeature(
      di,
      messagingFeature,
      messagingTestUtils.messagingFeatureForUnitTesting,
      routingFeature,
      loggerFeature,
      animateFeature,
      clusterSidebarFeature,
      randomFeature,
      kubeApiSpecificsFeature,
      notificationsFeature,
    );
  });

  di.preventSideEffects();

  runInAction(() => {
    const injectables = global.injectablePaths.renderer.paths
      .map((path) => require(path))
      .flatMap(Object.values)
      .filter(isInjectable);

    for (const block of chunk(100)(injectables)) {
      di.register(...block);
    }
  });

  for (const globalOverridePath of global.injectablePaths.renderer.globalOverridePaths) {
    const globalOverride = require(globalOverridePath).default as GlobalOverride<unknown, unknown, unknown>;

    di.override(globalOverride.injectable, globalOverride.overridingInstantiate);
  }

  [startTopbarStateSyncInjectable].forEach((injectable) => {
    di.override(injectable, () => ({
      id: injectable.id,
      run: () => {},
    }));
  });

  di.override(terminalSpawningPoolInjectable, () => document.createElement("div"));
  di.override(hostedClusterIdInjectable, () => undefined);

  di.override(legacyOnChannelListenInjectable, () => () => noop);

  di.override(requestAnimationFrameInjectable, () => (callback) => callback());
  di.override(watchHistoryStateInjectable, () => () => () => {});

  getOverrideFsWithFakes()(di);

  return di;
};
