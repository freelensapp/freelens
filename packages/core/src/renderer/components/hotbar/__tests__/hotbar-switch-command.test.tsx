import type { RenderResult } from "@testing-library/react";
import type { IComputedValue } from "mobx";
import React from "react";
import directoryForUserDataInjectable from "../../../../common/app-paths/directory-for-user-data/directory-for-user-data.injectable";
import createHotbarInjectable, {
  type CreateHotbar,
} from "../../../../features/hotbar/storage/common/create-hotbar.injectable";
import type { Hotbar } from "../../../../features/hotbar/storage/common/hotbar";
import hotbarsInjectable from "../../../../features/hotbar/storage/common/hotbars.injectable";
import hotbarsStateInjectable from "../../../../features/hotbar/storage/common/state.injectable";
import { getDiForUnitTesting } from "../../../getDiForUnitTesting";
import { renderFor } from "../../test-utils/renderFor";
import { HotbarSwitchCommand } from "../hotbar-switch-command";

describe("when there is one hotbar", () => {
  let result: RenderResult;
  let createHotbar: CreateHotbar;
  let hotbars: IComputedValue<Hotbar[]>;

  beforeEach(() => {
    const di = getDiForUnitTesting();
    const render = renderFor(di);

    di.override(directoryForUserDataInjectable, () => "/some-directory-for-user-data");

    createHotbar = di.inject(createHotbarInjectable);
    hotbars = di.inject(hotbarsInjectable);

    const hotbarsState = di.inject(hotbarsStateInjectable);
    const defaultHotbar = createHotbar({ name: "default" });

    hotbarsState.set(defaultHotbar.id, defaultHotbar);

    result = render(<HotbarSwitchCommand />);
  });

  it("renders w/o errors", () => {
    expect(result.container).toMatchSnapshot();
  });

  it("'Remove hotbar ...' must not be present", () => {
    expect(hotbars.get().length).toBe(1);
    expect(result.queryByText("Remove hotbar ...")).not.toBeInTheDocument();
  });
});

describe("when there are two hotbars", () => {
  let result: RenderResult;
  let createHotbar: CreateHotbar;
  let hotbars: IComputedValue<Hotbar[]>;

  beforeEach(() => {
    const di = getDiForUnitTesting();
    const render = renderFor(di);

    di.override(directoryForUserDataInjectable, () => "/some-directory-for-user-data");

    createHotbar = di.inject(createHotbarInjectable);
    hotbars = di.inject(hotbarsInjectable);

    const hotbarsState = di.inject(hotbarsStateInjectable);
    const defaultHotbar = createHotbar({ name: "default" });
    const nonDefaultHotbar = createHotbar({ name: "non-default" });

    hotbarsState.set(defaultHotbar.id, defaultHotbar);
    hotbarsState.set(nonDefaultHotbar.id, nonDefaultHotbar);

    result = render(<HotbarSwitchCommand />);
  });

  it("renders w/o errors", () => {
    expect(result.container).toMatchSnapshot();
  });

  it("'Remove hotbar ...' must be present", () => {
    expect(hotbars.get().length).toBe(2);
    expect(result.queryByText("Remove hotbar ...")).toBeInTheDocument();
  });
});
