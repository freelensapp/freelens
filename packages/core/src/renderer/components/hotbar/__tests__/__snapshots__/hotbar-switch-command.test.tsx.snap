// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when there are two hotbars renders w/o errors 1`] = `
<div>
  <div
    class="Select theme-dark css-b62m3t-container"
  >
    <span
      class="css-1f43avz-a11yText-A11yText"
      id="react-select-switch-to-hotbar-input-live-region"
    >
      <span
        id="aria-selection"
      >
        option , selected.
      </span>
      <span
        id="aria-focused"
      />
      <span
        id="aria-results"
      >
        5 results available.
      </span>
      <span
        id="aria-guidance"
      >
        Use Up and Down to choose options, press Enter to select the currently focused option, press Escape to exit the menu, press Tab to select the option and exit the menu.
      </span>
    </span>
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      class="css-1f43avz-a11yText-A11yText"
      role="log"
    />
    <div
      class="Select__control Select__control--is-focused Select__control--menu-is-open css-t3ipsp-control"
    >
      <div
        class="Select__value-container css-hlgwow"
      >
        <div
          class="Select__placeholder css-1jqq78o-placeholder"
          id="react-select-switch-to-hotbar-input-placeholder"
        >
          Switch to hotbar
        </div>
        <div
          class="Select__input-container css-19bb58m"
          data-value=""
        >
          <input
            aria-activedescendant=""
            aria-autocomplete="list"
            aria-controls="react-select-switch-to-hotbar-input-listbox"
            aria-describedby="react-select-switch-to-hotbar-input-placeholder"
            aria-expanded="true"
            aria-haspopup="true"
            autocapitalize="none"
            autocomplete="off"
            autocorrect="off"
            class="Select__input"
            id="switch-to-hotbar-input"
            role="combobox"
            spellcheck="false"
            style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
            tabindex="0"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="Select__indicators css-1wy0on6"
      />
    </div>
    <div
      class="theme-dark switch-to-hotbar-input-options Select__menu css-1nmdiq5-menu"
    >
      <div
        aria-multiselectable="false"
        class="Select__menu-list css-qr46ko"
        id="react-select-switch-to-hotbar-input-listbox"
        role="listbox"
      >
        <div
          aria-disabled="false"
          aria-selected="false"
          class="Select__option css-10wo9uf-option"
          id="react-select-switch-to-hotbar-input-option-0"
          role="option"
          tabindex="-1"
        >
          1: default
        </div>
        <div
          aria-disabled="false"
          aria-selected="false"
          class="Select__option css-10wo9uf-option"
          id="react-select-switch-to-hotbar-input-option-1"
          role="option"
          tabindex="-1"
        >
          2: non-default
        </div>
        <div
          aria-disabled="false"
          aria-selected="false"
          class="Select__option css-10wo9uf-option"
          id="react-select-switch-to-hotbar-input-option-2"
          role="option"
          tabindex="-1"
        >
          Add hotbar ...
        </div>
        <div
          aria-disabled="false"
          aria-selected="false"
          class="Select__option css-10wo9uf-option"
          id="react-select-switch-to-hotbar-input-option-3"
          role="option"
          tabindex="-1"
        >
          Remove hotbar ...
        </div>
        <div
          aria-disabled="false"
          aria-selected="false"
          class="Select__option css-10wo9uf-option"
          id="react-select-switch-to-hotbar-input-option-4"
          role="option"
          tabindex="-1"
        >
          Rename hotbar ...
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`when there is one hotbar renders w/o errors 1`] = `
<div>
  <div
    class="Select theme-dark css-b62m3t-container"
  >
    <span
      class="css-1f43avz-a11yText-A11yText"
      id="react-select-switch-to-hotbar-input-live-region"
    >
      <span
        id="aria-selection"
      >
        option , selected.
      </span>
      <span
        id="aria-focused"
      />
      <span
        id="aria-results"
      >
        3 results available.
      </span>
      <span
        id="aria-guidance"
      >
        Use Up and Down to choose options, press Enter to select the currently focused option, press Escape to exit the menu, press Tab to select the option and exit the menu.
      </span>
    </span>
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      class="css-1f43avz-a11yText-A11yText"
      role="log"
    />
    <div
      class="Select__control Select__control--is-focused Select__control--menu-is-open css-t3ipsp-control"
    >
      <div
        class="Select__value-container css-hlgwow"
      >
        <div
          class="Select__placeholder css-1jqq78o-placeholder"
          id="react-select-switch-to-hotbar-input-placeholder"
        >
          Switch to hotbar
        </div>
        <div
          class="Select__input-container css-19bb58m"
          data-value=""
        >
          <input
            aria-activedescendant=""
            aria-autocomplete="list"
            aria-controls="react-select-switch-to-hotbar-input-listbox"
            aria-describedby="react-select-switch-to-hotbar-input-placeholder"
            aria-expanded="true"
            aria-haspopup="true"
            autocapitalize="none"
            autocomplete="off"
            autocorrect="off"
            class="Select__input"
            id="switch-to-hotbar-input"
            role="combobox"
            spellcheck="false"
            style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
            tabindex="0"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="Select__indicators css-1wy0on6"
      />
    </div>
    <div
      class="theme-dark switch-to-hotbar-input-options Select__menu css-1nmdiq5-menu"
    >
      <div
        aria-multiselectable="false"
        class="Select__menu-list css-qr46ko"
        id="react-select-switch-to-hotbar-input-listbox"
        role="listbox"
      >
        <div
          aria-disabled="false"
          aria-selected="false"
          class="Select__option css-10wo9uf-option"
          id="react-select-switch-to-hotbar-input-option-0"
          role="option"
          tabindex="-1"
        >
          1: default
        </div>
        <div
          aria-disabled="false"
          aria-selected="false"
          class="Select__option css-10wo9uf-option"
          id="react-select-switch-to-hotbar-input-option-1"
          role="option"
          tabindex="-1"
        >
          Add hotbar ...
        </div>
        <div
          aria-disabled="false"
          aria-selected="false"
          class="Select__option css-10wo9uf-option"
          id="react-select-switch-to-hotbar-input-option-2"
          role="option"
          tabindex="-1"
        >
          Rename hotbar ...
        </div>
      </div>
    </div>
  </div>
</div>
`;
